nntrainer/layers/pooling2d_layer.cpp

```diff
@@ -139,18 +163,30 @@ void Pooling2DLayer::calcDerivative() {
       }
     }
   } break;
+  case PoolingType::global_average:
   case PoolingType::average: {
+    int heigth_stride_end = height - p_height + padding[0];
+    int width_stride_end = width - p_width + padding[1];
+    auto iter = max_idx.begin();
     for (unsigned int b = 0; b < batch; ++b) {
       for (unsigned int i = 0; i < channel; ++i) {
         J = 0;
-        for (unsigned int j = 0; j <= height - p_height; j += stride[0]) {
+        for (int j = -padding[0]; j <= heigth_stride_end; j += stride[0]) {
           K = 0;
-          for (unsigned int k = 0; k <= width - p_width; k += stride[1]) {
-            float del = deriv.getValue(b, i, J, K) / static_cast<float>(p_size);
-            for (unsigned int pi = 0; pi < p_height; ++pi) {
-              for (unsigned int pj = 0; pj < p_width; ++pj) {
-                result.setValue(b, i, j + pi, k + pj,
-                                result.getValue(b, i, j + pi, k + pj) + del);
+          for (int k = -padding[1]; k <= width_stride_end; k += stride[1]) {
+            float del = deriv.getValue(b, i, J, K) / *iter;
+            iter++;
+            int patch_height_end = j + p_height;
+            int patch_width_end = k + p_width;
+            for (int h = j; h < patch_height_end; ++h) {
+              if (h < 0 || height <= h) {
+                continue;
+              }
+              for (int w = k; w < patch_width_end; ++w) {
```

Comment:
please merge the conditions below in the loop itself.

Reply:
This can be done at a lot of places. Please update them all.

Reply:
Yes, I think it's good idea